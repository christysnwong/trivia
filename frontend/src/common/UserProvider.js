import React, {useEffect, useState} from "react";
import UserContext from "./UserContext";

const demoUser = {
  id: 1,
  username: "testuser",
  first_name: "Test",
  last_name: "User",
  email: "test@test.com",
  stats: {
    level: 14,
    title: "Pro",
    "quizzes completed": 34,
    points: 3440,
    remainingPts: 60,
    levelPts: 300,
  },
};

const UserProvider = ({
  children,
  currUser = demoUser
}) => {
  
  const [currentUser, setCurrUser] = useState();

  return (
    // <UserContext.Provider value={{ currUser }}>
    <UserContext.Provider value={{ currUser, setCurrUser }}>
      {children}
    </UserContext.Provider>
  );
};

export default UserProvider;
