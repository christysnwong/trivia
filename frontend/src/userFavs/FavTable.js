import {
  Button,
  ButtonGroup,
  UncontrolledDropdown,
  DropdownToggle,
  DropdownMenu,
  DropdownItem,
} from "reactstrap";


const FavTable = ({ userFolders, folder, entries, removeTrivia, moveTrivia }) => {
  return (
    <div>

      {folder.name && (
        <h3 className="mb-4">
          <i className="fa-solid fa-folder-open me-2"></i>
          {folder.name}
        </h3>
      )}
      {!folder.name && (
        <>
          <p>
            You can either create a new folder by clicking on{" "}
            <i className="fa-solid fa-folder-plus"></i> or select a folder above
            to view your favourited trivia.
          </p>
        </>
      )}
      {entries.length === 0 && folder.name && "This folder is empty."}
      {entries.length > 0 && (
        <div>
          <table className="table table-striped table-sm">
            <thead>
              <tr>
                <th scope="col">#</th>
                {!!entries.length &&
                  Object.keys(entries[0])
                    .filter((heading) => heading !== "id")
                    .map((heading, idx) => (
                      <th
                        scope="col"
                        key={idx}
                        className="text-capitalize fw-bold"
                      >
                        {heading}
                      </th>
                    ))}
                <th scope="col" className="text-capitalize fw-bold d-none d-sm-table-cell">
                  Options
                </th>
              </tr>
            </thead>
            <tbody>
              {entries.map((entry, idx) => (
                <tr key={idx} id={entry.id}>
                  <th scope="row">{idx + 1}</th>

                  {Object.values(entry)
                    .filter((field) => typeof field !== "number")
                    .map((field, idx) => (
                      <td key={idx} className="text-capitalize">
                        <span
                          dangerouslySetInnerHTML={{ __html: field }}
                        ></span>
                      </td>
                    ))}

                  <td className="d-none d-sm-table-cell">
                    <ButtonGroup className="my-0" size="sm">
                      <ButtonGroup size="sm">
                        <UncontrolledDropdown>
                          <DropdownToggle caret size="sm">
                            <i className="fa-solid fa-up-down-left-right"></i>
                          </DropdownToggle>
                          <DropdownMenu>
                            <DropdownItem header>Move to...</DropdownItem>
                            <DropdownItem divider />
                            {userFolders.map((f) => (
                              <DropdownItem
                                key={f.folderId}
                                id={f.folderId}
                                onClick={moveTrivia}
                              >
                                {f.name}
                              </DropdownItem>
                            ))}
                          </DropdownMenu>
                        </UncontrolledDropdown>
                      </ButtonGroup>
                      <Button onClick={removeTrivia} color="danger">
                        <i className="fa-sharp fa-solid fa-trash"></i>
                      </Button>
                    </ButtonGroup>
                  </td>
                </tr>
              ))}
            </tbody>
          </table>
        </div>
      )}
    </div>
  );
};

export default FavTable;



