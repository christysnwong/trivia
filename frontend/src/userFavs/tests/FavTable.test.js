import React from "react";
import { render, screen } from "@testing-library/react";
import FavTable from "../FavTable";
import { MemoryRouter } from "react-router-dom";
import UserProvider from "../../common/UserProvider";

const userFolders = [
  {
    folderId: 1,
    name: "All",
  },
  {
    folderId: 2,
    name: "TEST1",
  },
  {
    folderId: 3,
    name: "TEST2",
  },
  {
    folderId: 4,
    name: "TEST3",
  },
];

const folder = { id: 2, name: "TEST1" };

const entries = [
  {
    id: 4,
    question: "U1 Question 1 in TEST1",
    answer: "Answer 1",
  },
];



it("renders without crashing", function () {
  render(
    <UserProvider>
      <FavTable userFolders={userFolders} folder={folder} entries={entries} />
    </UserProvider>
  );
});

it("matches snapshot", function () {
  const { asFragment } = render(
    <MemoryRouter>
      <UserProvider>
        <FavTable userFolders={userFolders} folder={folder} entries={entries}/>
      </UserProvider>
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});


it("loads the question and answer", function () {
  render(
    <UserProvider>
      <FavTable userFolders={userFolders} folder={folder} entries={entries} />
    </UserProvider>
  );

  expect(screen.getAllByText("TEST1")).toHaveLength(2);
  expect(screen.getByText("U1 Question 1 in TEST1")).toBeInTheDocument();
  expect(screen.getByText("Answer 1")).toBeInTheDocument();

  expect(screen.getByText("Move to...")).toBeInTheDocument();


})