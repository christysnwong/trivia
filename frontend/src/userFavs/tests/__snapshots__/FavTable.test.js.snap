// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h3
      class="mb-4"
    >
      <i
        class="fa-solid fa-folder-open me-2"
      />
      TEST1
    </h3>
    <div>
      <table
        class="table table-striped table-sm"
      >
        <thead>
          <tr>
            <th
              scope="col"
            >
              #
            </th>
            <th
              class="text-capitalize fw-bold"
              scope="col"
            >
              question
            </th>
            <th
              class="text-capitalize fw-bold"
              scope="col"
            >
              answer
            </th>
            <th
              class="text-capitalize fw-bold d-none d-sm-table-cell"
              scope="col"
            >
              Options
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            id="4"
          >
            <th
              scope="row"
            >
              1
            </th>
            <td
              class="text-capitalize"
            >
              <span>
                U1 Question 1 in TEST1
              </span>
            </td>
            <td
              class="text-capitalize"
            >
              <span>
                Answer 1
              </span>
            </td>
            <td
              class="d-none d-sm-table-cell"
            >
              <div
                class="my-0 btn-group-sm btn-group"
                role="group"
              >
                <div
                  class="btn-group-sm btn-group"
                  role="group"
                >
                  <div
                    class="dropdown"
                  >
                    <button
                      aria-expanded="false"
                      aria-haspopup="true"
                      class="dropdown-toggle btn btn-secondary btn-sm"
                      type="button"
                    >
                      <i
                        class="fa-solid fa-up-down-left-right"
                      />
                    </button>
                    <div
                      aria-hidden="true"
                      class="dropdown-menu"
                      role="menu"
                      tabindex="-1"
                    >
                      <h6
                        class="dropdown-header"
                        tabindex="-1"
                      >
                        Move to...
                      </h6>
                      <div
                        class="dropdown-divider"
                        tabindex="-1"
                      />
                      <button
                        class="dropdown-item"
                        id="1"
                        role="menuitem"
                        tabindex="0"
                        type="button"
                      >
                        All
                      </button>
                      <button
                        class="dropdown-item"
                        id="2"
                        role="menuitem"
                        tabindex="0"
                        type="button"
                      >
                        TEST1
                      </button>
                      <button
                        class="dropdown-item"
                        id="3"
                        role="menuitem"
                        tabindex="0"
                        type="button"
                      >
                        TEST2
                      </button>
                      <button
                        class="dropdown-item"
                        id="4"
                        role="menuitem"
                        tabindex="0"
                        type="button"
                      >
                        TEST3
                      </button>
                    </div>
                  </div>
                </div>
                <button
                  class="btn btn-danger"
                >
                  <i
                    class="fa-sharp fa-solid fa-trash"
                  />
                </button>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</DocumentFragment>
`;
